Algorithm.Framework/Alphas/HistoricalReturnsAlphaModel.cs

```diff
@@ -0,0 +1,148 @@
+            Resolution resolution = Resolution.Daily
+            )
+        {
+            _lookback = lookback;
+            _resolution = resolution;
+            _predictionInterval = _resolution.ToTimeSpan().Multiply(_lookback);
+            _symbolDataBySymbol = new Dictionary<Symbol, SymbolData>();
+            Name = $"{nameof(HistoricalReturnsAlphaModel)}({lookback},{resolution})";
+        }
+
+        /// <summary>
+        /// Updates this alpha model with the latest data from the algorithm.
+        /// This is called each time the algorithm receives data for subscribed securities
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance</param>
+        /// <param name="data">The new data available</param>
+        /// <returns>The new insights generated</returns>
+        public IEnumerable<Insight> Update(QCAlgorithmFramework algorithm, Slice data)
+        {
+            var insights = new List<Insight>();
+            foreach (var symbolData in _symbolDataBySymbol.Values)
+            {
+                if (symbolData.CanEmit())
+                {
+                    var direction = InsightDirection.Flat;
+                    var magnitude = (double)symbolData.ROC.Current.Value;
+                    if (magnitude > 0) direction = InsightDirection.Up;
+                    if (magnitude < 0) direction = InsightDirection.Down;
+                    insights.Add(Insight.Price(symbolData.Security.Symbol, _predictionInterval, direction, magnitude, null));
+                }
+            }
+            return insights;
+        }
+
+        /// <summary>
+        /// Event fired each time the we add/remove securities from the data feed
+        /// </summary>
+        /// <param name="algorithm">The algorithm instance that experienced the change in securities</param>
+        /// <param name="changes">The security additions and removals from the algorithm</param>
+        public void OnSecuritiesChanged(QCAlgorithmFramework algorithm, SecurityChanges changes)
+        {
+            // clean up data for removed securities
+            foreach (var removed in changes.RemovedSecurities)
+            {
+                SymbolData data;
+                if (_symbolDataBySymbol.TryGetValue(removed.Symbol, out data))
+                {
+                    _symbolDataBySymbol.Remove(removed.Symbol);
+                    algorithm.SubscriptionManager.RemoveConsolidator(removed.Symbol, data.Consolidator);
+                }
+            }
+
+            // initialize data for added securities
+            foreach (var added in changes.AddedSecurities)
+            {
+                if (!_symbolDataBySymbol.ContainsKey(added.Symbol))
+                {
+                    var symbolData = new SymbolData(algorithm, added, _lookback, _resolution);
+                    _symbolDataBySymbol.Add(added.Symbol, symbolData);
+                    symbolData.WarmUpIndicators(algorithm, added, _lookback, _resolution);
```

Comment:
I can see that you tried to mimic the python version, but please take a look at how indicators were warmed up in [RsilAlphaModel](https://github.com/QuantConnect/Lean/blob/master/Algorithm.Framework/Alphas/RsiAlphaModel.cs#L131), since this is the pattern we should use in C#.